### PR TITLE
top10 for requeue & timeout

### PR DESCRIPTION
**Pull request by DoraALin**
for issue #39  Signed-off-by: 元守 <lulin@youzan.com>

### CODE REVIEW

File: nsqadmin/http.go [hunk 1]
```diff
@@ -1036,24 +1036,65 @@ type counterStats struct {
 }
 
 type rankStats struct {
-	TopicName         string `json:"topic_name"`
-	TotalChannelDepth int64  `json:"total_channel_depth"`
-	MessageCount      int64  `json:"message_count"`
-	HourlyPubSize     int64  `json:"hourly_pubsize"`
+	Name         string 	 `json:"name"`
```
Comment (absolute8511) [inline nsqadmin/http.go:1039]: Why change the `topic name` to `name`

File: nsqadmin/http.go [hunk 2]
```diff
@@ -1103,7 +1144,7 @@ func (s *httpServer) statisticsHandler(w http.ResponseWriter, req *http.Request,
 	filter := ps.ByName("filter")
 	s.ctx.nsqadmin.logf("filter passed in statisticsHandler: " + filter)
 	if "" == filter {
-		filter_str := []string{"channel-depth", "hourly-pubsize"}
+		filter_str := []string{"channel-depth", "hourly-pubsize", "channel-timeout", "channel-requeue"}
```
Comment (absolute8511) [inline nsqadmin/http.go]: rename to `sortByStr` and avoid underline in Golang variable name.

File: nsqadmin/http.go [hunk 3]
```diff
@@ -1103,7 +1144,7 @@ func (s *httpServer) statisticsHandler(w http.ResponseWriter, req *http.Request,
 	filter := ps.ByName("filter")
```
Comment (absolute8511) [inline nsqadmin/http.go]: Maybe rename it to `sortBy` ?

File: nsqadmin/http.go [hunk 4]
```diff
@@ -1139,6 +1184,51 @@ func (s *httpServer) statisticsHandler(w http.ResponseWriter, req *http.Request,
 		messages = append(messages, pe.Error())
 	}
 
+	if filter == "channel-timeout" || filter == "channel-requeue" {
```
Comment (absolute8511) [inline nsqadmin/http.go]: Why not move sort code to the `GetNSQDStats` just like others and only handle removing the 0 time/req here.

File: nsqadmin/http.go [hunk 5]
```diff
@@ -1113,6 +1154,10 @@ func (s *httpServer) statisticsHandler(w http.ResponseWriter, req *http.Request,
 	switch filter {
 	case "channel-depth":
 		rankName = "Top10 topics in Total Channel Depth"
+	case "channel-timeout":
```
Comment (absolute8511) [inline nsqadmin/http.go]: Maybe add test case for these sort statistics

File: nsqadmin/http.go (hunk 1 above)
Comment (DoraALin) [reply nsqadmin/http.go:1039]: as rank of channel timeout&requeue added. I woule like to give a more general name.

File: nsqadmin/http.go (hunk 3 above)
Comment (DoraALin) [reply nsqadmin/http.go]: will do.

File: nsqadmin/http.go (hunk 2 above)
Comment (DoraALin) [reply nsqadmin/http.go]: will do

File: nsqadmin/http.go (hunk 4 above)
Comment (DoraALin) [reply nsqadmin/http.go]: as channel stats returns as map, and sort performed against topic stats list inside GetNSQStats, hence I sort reference list to channel stats map after channel stats map returned.

Comment (absolute8511) [reply nsqadmin/http.go]: Can we sort just like the channel depth to make it easier? We can merge the timeout/req under the same topic so we can handle it more graceful.

**Comment by DoraALin** [comment]
@absolute8511 for your review. I moved the logic about sort(channel timeout& requeue) into statistcsHandler, rename filter to sortBy

**Review by absolute8511 — APPROVED** [review]
LGTM

**Pull request merged**